### PR TITLE
refactor: Update React dependencies

### PR DESCRIPTION
**Pull request by alexvy86**
## Description

Update React from 17 to 18 in our examples and in Devtools. Also cleans up the `@types/react-dom` dev dependency from packages that don't import `react-dom`, and adds `@types/react` and `@types/react-dom` as dev dependencies to a package that uses both of them but didn't have a dependency on the types.

This is motivated by work on the Fluid AI Collab experiment which uses a NextJS app as an example (because it wants to run some code server side) and NextJS requires React 18.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

I confirmed that the Devtools example app, the clicker example, and the bubblebench-shared-tree example all worked as expected.

### CODE REVIEW

File: examples/apps/data-object-grid/package.json [hunk 1]
```diff
@@ -56,10 +56,10 @@
 		"@fluidframework/runtime-definitions": "workspace:~",
 		"@fluidframework/runtime-utils": "workspace:~",
 		"prop-types": "^15.8.1",
-		"react": "^17.0.1",
+		"react": "^18.3.1",
 		"react-collapsible": "^2.7.0",
-		"react-dom": "^17.0.1",
-		"react-grid-layout": "^0.18.3",
+		"react-dom": "^18.3.1",
+		"react-grid-layout": "^1.4.4",
```
Comment (alexvy86) [inline examples/apps/data-object-grid/package.json:62]: Went through the [changelog](https://github.com/react-grid-layout/react-grid-layout/blob/master/CHANGELOG.md) and the breaking changes don't seem to affect us.

File: examples/data-objects/multiview/interface/package.json [hunk 2]
```diff
@@ -47,7 +47,6 @@
 		"@fluidframework/build-common": "^2.0.3",
 		"@fluidframework/build-tools": "^0.46.0",
 		"@fluidframework/eslint-config-fluid": "^5.4.0",
-		"@types/react": "^17.0.44",
```
Comment (alexvy86) [inline examples/data-objects/multiview/interface/package.json:50]: This package doesn't depend on `react`.

**Review by Abe27342 — APPROVED** [review]
looks fine. Is the plan to update to idiomatic react 18 usage afterward? (e.g. [using createRoot over ReactDOM.render](https://react.dev/blog/2022/03/08/react-18-upgrade-guide#updates-to-client-rendering-apis))

**Review by seanimam — APPROVED** [review]

File: examples/benchmarks/bubblebench/baseline/.eslintrc.cjs [hunk 3]
```diff
@@ -5,5 +5,8 @@
 
 module.exports = {
 	extends: [require.resolve("@fluidframework/eslint-config-fluid"), "prettier"],
-	rules: {},
+	rules: {
+		// AB#18875
```
Comment (Josmithr) [inline examples/benchmarks/bubblebench/baseline/.eslintrc.cjs:9]: Can you also add an explanation of the TODO, in addition to the work item number? That way folks without access to our azure devops still have context.

File: package.json [hunk 4]
```diff
@@ -384,12 +384,13 @@
 			"@types/node is ignored because it is usually not needed by packages, and if it is, then the package will hit a compilation failure.",
 			"Peer dependencies on fluid-framework are never fulfilled since that's an in-repo dependency; we expect a customer using some packages to install it as a dependency.",
 			"markdown-magic is provided by another dependency so is ignored here.",
-			"oclif includes some AWS-related features, but we don't use them, so we ignore @aws-sdk peer dependencies."
+			"oclif includes some AWS-related features, but we don't use them, so we ignore @aws-sdk peer dependencies.",
+			"The react-split-pane package used by devtools-view has a peer dependency on React 16, but it doesn't seem to be maintained and it works fine with React 18."
```
Comment (Josmithr) [inline package.json:388]: Can we file a work item to look into alternatives to that library? If it isn't maintained, hopefully we can find something that is.

**Review by Josmithr — APPROVED** [review]

File: examples/benchmarks/bubblebench/baseline/.eslintrc.cjs (hunk 3 above)
Comment (Josmithr) [reply examples/benchmarks/bubblebench/baseline/.eslintrc.cjs:9]: We should also probably prefix these with "TODO" for clarity.

**Review by CraigMacomber — APPROVED** [review]

**Pull request merged**